experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -337,21 +340,42 @@ std::optional<Value> findControlInputToBB(Operation *op) {
   handshake::FuncOp funcOp = op->getParentOfType<handshake::FuncOp>();
   assert(funcOp && "op should have parent function");
 
-  // Find input arcs to BBs in the IR
-  BBtoArcsMap bbToPredecessorArcs = getBBPredecessorArcs(funcOp);
-  unsigned bb = getLogicBB(op).value();
+  std::optional<unsigned> targetBB = getLogicBB(op);
+  if (!targetBB) {
+    op->emitError("Operation does not have a BB.");
+    return {};
+  }
+
+  // We use the control token, which is an input to the control branch
+  // as the enable signal for the speculator.
+  mlir::Value ctrlSignal;
+  bool isControlBranchFound = false;
+  for (auto branchOp : funcOp.getOps<handshake::ConditionalBranchOp>()) {
+    // Check if the branch is in the same BB as the operation
+    // specified as the location for the speculator
+    if (auto brBB = getLogicBB(branchOp);
```

Comment:
previously you had this complicated safety check for getLogicBB and here you use it raw. is the complicated safety check above actually needed?

Reply:
The previous one is this. The main difference appears to be whether we specify the exact type `std::optional<...>` or not. What do you mean by the complicated safety check?

```cpp
  std::optional<unsigned> targetBB = getLogicBB(op);
  if (!targetBB) {
    op->emitError("Operation does not have a BB.");
    return {};
  }
```

Reply:
yeah, is this necessary? it seems not to be if you didn't do it on line 356

Reply:
I want to ensure the returned value is valid unless the return type of getLogicBB can be changed from `std::optional<unsigned>` to `unsigned`. Although C++ is not a type-safe language, I want to prioritize type safety as much as possible.
Note that this check is conducted in various methods both inside and outside the speculation pass, like `findControlBranch`, `PlacementFinder::findCommitPositions`, `PlacementFinder::findSaveCommitsTraversal`, `getBlockControl`, and `MemoryInterfaceBuilder::determineInterfaceInputs`.

Reply:
Possibly I might be biased because I like functional languages more 🙂

Reply:
Yeah I fully see why to do it, but either we do it always or not at all: which should it be?